reccmp/tools/project.py

```diff
@@ -0,0 +1,457 @@
+    return 0
+
+
+class TargetType(enum.Enum):
+    SHARED_LIBRARY = "SHARED_LIBRARY"
+    EXECUTABLE = "EXECUTABLE"
+
+
+def executable_or_library(path: Path) -> TargetType:
+    str_path = str(path).lower()
+    if str_path.endswith(".dll"):
+        return TargetType.SHARED_LIBRARY
+    if str_path.endswith(".exe"):
+        return TargetType.EXECUTABLE
+    # FIXME: detect from file contents (or arguments?)
+    raise ValueError("Unknown target type")
+
+
+def get_default_cmakelists_txt(project_name: str, targets: dict[str, Path]) -> str:
+    result = textwrap.dedent(
+        f"""\
+        cmake_minimum_required(VERSION 3.20)
+        project({project_name})
+
+        include("${{CMAKE_CURRENT_SOURCE_DIR}}/cmake/reccmp.cmake")
+    """
+    )
+
+    for target_name, target_path in targets.items():
+        target_type = executable_or_library(target_path)
+        target_prefix = ""
+        target_suffix = target_path.suffix
+        if target_type == TargetType.SHARED_LIBRARY and target_name.startswith("lib"):
+            target_prefix = "lib"
+            target_name = target_name.removeprefix("lib")
+
+        match target_type:
+            case TargetType.EXECUTABLE:
+                add_executable_or_library = "add_executable"
+                maybe_shared = ""
+            case TargetType.SHARED_LIBRARY:
+                add_executable_or_library = "add_library"
+                maybe_shared = "SHARED"
+        result += "\n"
+        result += textwrap.dedent(
+            f"""\
+            {add_executable_or_library}({target_name} {maybe_shared}
+                main_{target_name}.cpp
+                main_{target_name}.hpp
+            )
+            reccmp_add_target({target_name} ID {target_name})
+            set_property(TARGET {target_name} PROPERTY OUTPUT_NAME "{target_path.stem}")
+            set_property(TARGET {target_name} PROPERTY PREFIX "{target_prefix}")
+            set_property(TARGET {target_name} PROPERTY SUFFIX "{target_suffix}")
+        """
+        )
+
+    result += "\n"
+    result += textwrap.dedent(
+        """\
+        reccmp_configure()
+    """
+    )
+    return result
+
+
+def get_default_main_hpp(target_id: str) -> str:
+    return textwrap.dedent(
+        f"""\
+        #ifndef {target_id.upper()}_HPP
+        #define {target_id.upper()}_HPP
+
+        // VTABLE: {target_id} 0x10001000
+        // SIZE 0x8
+        class SomeClass {{
+            virtual ~SomeClass(); // vtable+0x00
+            int m_member;
+        }};
+
+        #endif /* {target_id.upper()}_HPP */
+        """
+    )
+
+
+def get_default_main_cpp(target_id: str, original_path: Path, hpp_path: Path) -> str:
+    target_type = executable_or_library(original_path)
+    match target_type:
+        case TargetType.EXECUTABLE:
+            entry_function = textwrap.dedent(
+                """\
+                #ifdef _WIN32
+                #include <windows.h>
+
+                // FUNCTION: {original_id} 0x10000020
```

Comment:
Should these both be f-strings with `{target_id.upper()}` to match `get_default_main_hpp()` ?

Reply:
The code does not use `original_id` anymore, but uses `target_id` instead.
(I also moved the project modification/generation out of the tool into a proper module, adding unit tests)